e2e/libs/utility/utility.py

```diff
@@ -1,31 +1,131 @@
-from kubernetes import config, client
-from longhorn import from_env
+import os
+import requests
+import warnings
 import string
 import random
-import os
+import time
+import logging
+
+from . import globalVars
+from kubernetes import config, client
+from kubernetes.client import configuration
+from longhorn import from_env
+from pick import pick
+
+MAX_SUPPORT_BINDLE_NUMBER = 20
+RETRY_EXEC_COUNTS = 30
+RETRY_INTERVAL_LONG = 2
+
 
 class Utility:
 
+    @classmethod
     def generate_volume_name(cls):
         return "vol-" + \
             ''.join(random.choice(string.ascii_lowercase + string.digits)
                     for _ in range(6))
 
+    @classmethod
     def init_k8s_api_client(cls):
-        print('init_k8s_api_client')
+        logging.info('initiate the K8s api clients')
+
         # to make it easier to develop and debug
         # run test in local environment instead of in-cluster
         # our existing test is running in in-cluster environment
         # it makes us always need to build new docker image to test our code
-        config.load_kube_config()  # for local environment
-        # config.load_incluster_config()  # for in-cluster environment
+        contexts, active_context = config.list_kube_config_contexts(
```

Comment:
This is because the KubeConfig file may contain multiple contexts, which are loaded for selection like below:
![image](https://github.com/longhorn/longhorn-tests/assets/17548901/488b7188-9b4c-4705-80d2-0c0bada63897)

Reply:
Currently I'm not sure if it's right decision to use `config.load_kube_config()` instead of `config.load_incluster_config()`.

If we use `config.load_incluster_config()` here, will it make things simpler?

Reply:
Using `config.load_incluster_config()` is a simple way but need user switch to target cluster manually first. Currently implementation way will allow user select target cluster to proceed testing, I think it will be more flexible.

Reply:
Yes, I think the user manually specify what cluster they'd like to operate and then call `config.load_incluster_config()` is more reasonable. Can we switch back to use `config.load_incluster_config()`. Sorry for the confusing.

Reply:
As mentioned in the code, I think there may be cases where users need to run tests in local environment due to development and debugging. So, is it ok if I update the function like below? Or just keep `config.load_incluster_config()`.

```python
# load_incluster_config() needs two environ variables: KUBERNETES_SERVICE_HOST & KUBERNETES_SERVICE_PORT
if "KUBERNETES_SERVICE_HOST" in os.environ and "KUBERNETES_SERVICE_PORT" in os.environ:
    kubecfg = config.load_incluster_config()
else:
    contexts, active_context = config.list_kube_config_contexts(
        globalVars.Config.get("KUBECONFIG_PATH"))
    if not contexts:
        raise Exception("cannot find any context in kube-config file.")

    contexts = [context['name'] for context in contexts]
    active_index = contexts.index(active_context['name'])
    option, _ = pick(contexts, title="Pick the context to load",
                    default_index=active_index)
    kubecfg = config.load_kube_config(context=option)
```

Reply:
Can you help to check how it works in our existing test code?

Reply:
I've referred to functions under `/manager/integration/tests/common.py`, it define functions for each k8s clients.
https://github.com/longhorn/longhorn-tests/blob/63e570404511120b4eb734bfdac4fe5d078ac68f/manager/integration/tests/common.py#L1187-L1199

Therefore, I create these client creation functions in `utility.py` instead of in `globalVars.py`
https://github.com/longhorn/longhorn-tests/blob/7ba7fd8e586544ef2e9b78fd94d29291038ca14a/e2e/libs/utility/utility.py#L18-L24